File: payment-service/src/main/java/com/hedvig/paymentservice/serviceIntergration/productPricing/ProductPricingServiceImpl.java
package com.hedvig.paymentservice.serviceIntergration.productPricing;

import com.hedvig.paymentservice.query.member.entities.Transaction;
import com.hedvig.paymentservice.serviceIntergration.productPricing.dto.InsuranceStatus;
import com.hedvig.paymentservice.serviceIntergration.productPricing.dto.PolicyGuessRequestDto;
import com.hedvig.paymentservice.serviceIntergration.productPricing.dto.PolicyGuessResponseDto;
import feign.FeignException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;

import java.time.YearMonth;
import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static java.util.stream.Collectors.toList;

@Component
@Slf4j
public class ProductPricingServiceImpl implements ProductPricingService {
  private final ProductPricingClient client;

  public ProductPricingServiceImpl(final ProductPricingClient client) {
    this.client = client;
  }

  @Override
  public Optional<InsuranceStatus> getInsuranceStatus(String memberId) {
    try {
      ResponseEntity<InsuranceStatus> response = client.getInsuranceStatus(memberId);
      return response.getStatusCode().is2xxSuccessful() ? Optional.ofNullable(response.getBody()) : Optional.empty();
    } catch (FeignException ex) {
      switch (ex.status()) {
        case 500:
          log.error("Product-pricing returned 500 response");
        case 404:
          return Optional.empty();
      }
    }
    return Optional.empty();
  }

  @Override
  public Map<UUID, Optional<PolicyGuessResponseDto>> guessPolicyTypes(
    final Collection<Transaction> transactions,
    final YearMonth period
  ) {
    final Collection<PolicyGuessRequestDto> policyGuessDtos = transactions.stream()
      .map(PolicyGuessRequestDto::from)
      .collect(toList());
    return client.guessPolicyTypes(policyGuessDtos, period).getBody();
  }
}

File: payment-service/src/main/java/com/hedvig/paymentservice/serviceIntergration/productPricing/ProductPricingServiceImpl.kt
package com.hedvig.paymentservice.serviceIntergration.productPricing

import com.hedvig.paymentservice.query.member.entities.Transaction
import com.hedvig.paymentservice.serviceIntergration.productPricing.dto.ContractMarketInfo
import com.hedvig.paymentservice.serviceIntergration.productPricing.dto.ContractStateFilter
import com.hedvig.paymentservice.serviceIntergration.productPricing.dto.PolicyGuessRequestDto
import com.hedvig.paymentservice.serviceIntergration.productPricing.dto.PolicyGuessResponseDto
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.time.YearMonth
import java.util.Optional
import java.util.UUID
import java.util.stream.Collectors

@Component
class ProductPricingServiceImpl(
    private val client: ProductPricingClient
) : ProductPricingService {

    override fun guessPolicyTypes(
        transactions: Collection<Transaction>,
        period: YearMonth
    ): Map<UUID, Optional<PolicyGuessResponseDto>> {
        val policyGuessDtos: Collection<PolicyGuessRequestDto> = transactions.stream()
            .map { transaction: Transaction ->
                PolicyGuessRequestDto.from(transaction)
            }
            .collect(Collectors.toList())
        return client.guessPolicyTypes(policyGuessDtos, period).body!!
    }

    override fun getContractMarketInfo(memberId: String): ContractMarketInfo {
        return client.getContractMarketInfo(memberId).body!!
    }

    override fun hasContractActiveCurrentMonth(memberId: String): Boolean {
        val response = client.hasContract(memberId, ContractStateFilter.ACTIVE_CURRENT_MONTH)
        if (response.statusCode.is2xxSuccessful) {
            return response.body!!
        }
        return true
    }

    companion object {
        val logger = LoggerFactory.getLogger(this::class.java)!!
    }
}
